tests/test_e2e/test_cloud_spec.py

```diff
@@ -0,0 +1,48 @@
+import ops
+import pytest
+
+import scenario
+from scenario.state import CloudSpec
+
+
+@pytest.fixture(scope="function")
+def mycharm():
+    class MyCharm(ops.CharmBase):
+        def __init__(self, framework: ops.Framework):
+            super().__init__(framework)
+            for evt in self.on.events().values():
+                self.framework.observe(evt, self._on_event)
+
+        def _on_event(self, event):
+            pass
+
+    return MyCharm
+
+
+def test_get_cloud_spec(mycharm):
+    cloud_spec = CloudSpec.from_dict(
+        {
+            "name": "localhost",
+            "type": "lxd",
+            "endpoint": "https://127.0.0.1:8443",
+            "credential": {
+                "auth-type": "certificate",
+                "attrs": {
+                    "client-cert": "foo",
+                    "client-key": "bar",
+                    "server-cert": "baz",
+                },
+            },
+        }
+    )
+
+    ctx = scenario.Context(mycharm, meta={"name": "foo"})
+    with ctx.manager("start", scenario.State(cloud_spec=cloud_spec)) as mgr:
+        assert mgr.charm.model.get_cloud_spec() == cloud_spec
```

Comment:
Per previous discussion, I already ported the CloudSpec dataclass from ops to scenario. Why should it not work, since `scenario.state.CloudSpec` is exactly the same dataclass copied from `ops.model.CloudSpec` without changing anything?

I can change it to something like:

```python
def test_get_cloud_spec(mycharm):
    cloud_spec_dict = {
        "name": "localhost",
        "type": "lxd",
        "endpoint": "https://127.0.0.1:8443",
        "credential": {
            "auth-type": "certificate",
            "attrs": {
                "client-cert": "foo",
                "client-key": "bar",
                "server-cert": "baz",
            },
        },
    }
    scenario_cloud_spec = scenario.state.CloudSpec.from_dict(cloud_spec_dict)

    ctx = scenario.Context(mycharm, meta={"name": "foo"})
    with ctx.manager("start", scenario.State(cloud_spec=scenario_cloud_spec)) as mgr:
        assert mgr.charm.model.get_cloud_spec() == ops.CloudSpec.from_dict(
            cloud_spec_dict
        )
```

I'd argue this is messier, more confusing and error-prone. Maybe we should use ops.CloudSpec directly in scenario.state after all? At least in this use case (cloud spec), I don't see any value maintaining the same code in two places.

Reply:
> Per previous discussion, I already ported the CloudSpec dataclass from ops to scenario. Why should it not work, since `scenario.state.CloudSpec` is exactly the same dataclass copied from `ops.model.CloudSpec` without changing anything?

Because the `__eq__` that is provided by default in a dataclass compares the two objects as if they are tuples (which would work fine) but also requires that the two objects are the same type, which they are not.

```python
>>> @dataclasses.dataclass
... class A:
...   a: str = 'hello'
... 
>>> @dataclasses.dataclass
... class B:
...   a: str = 'hello'
... 
>>> a = A()
>>> b = B()
>>> a == b
False
```

`scenario.state.CloudSpec` has the same members (names and types) as `ops.model.CloudSpec` but is not the same type. You're creating the `cloud_spec` object just above, so it's clearly a `scenario.state.CloudSpec`, but `model.get_cloud_spec` needs to return an `ops.model.CloudSpec`, so something must be wrong here, because this test shouldn't pass as-is.

Reply:
This is definitely a bug, and needs to be fixed before this could be merged.

I dug a little and can see where the issue is - it's in mocking.py where the state's `cloud_spec` is returned. That needs to be converted to an `ops.CloudSpec` object. Similar to [this code that does it for notices](https://github.com/tonyandrewmeyer/ops-scenario/blob/8161af7408dd8bb674b17dfab8392f5b85df716a/scenario/mocking.py#L670).

Reply:
yeah there's more objects in state that have a `_to_ops_counterpart` utility method for mocking.py to use

Reply:
Sorry I have fixed it, resolved it without commented on it. It has been fixed in previous commits already.